recurring_contract/models/move.py

```diff
@@ -178,8 +178,102 @@ def find_sum(numbers, target, partial=None):
+        :param ref: The reference of the invoice, defaults to False (optional)
+        :param pay_mode_id: The payment mode to be used for the invoice, defaults to False (optional)
+        :return: A dictionary with the invoice_line_ids, date, payment_reference, and payment_mode_id.
+        """
+        # Ensure that the funciton receive one invoice
+        self.ensure_one()
+        # Build the dictionnary
+        inv_val_dict = {}
+        if contract:
+            inv_val_dict["invoice_line_ids"] = self._build_invoice_lines_from_contract(contract)
+        if due_date:
+            inv_val_dict["date"] = due_date
+        if ref:
+            inv_val_dict["payment_reference"] = ref
+        if pay_mode_id:
+            inv_val_dict["payment_mode_id"] = pay_mode_id
+        return {self.name: inv_val_dict}
+
+    def _build_invoice_lines_from_contract(self, contract):
+        """
+        It creates a list of tuples that will be used to create, modify or delete invoice lines.
+
+        :param contract recurring_contract that has some modification
+        """
+        res = []
+        line_ids = self.mapped("invoice_line_ids")
+        # Line to delete in the invoice
+        # If the invoice has different product than the contract we should delete the line
+        diff_product_invoices = [p for p in line_ids.mapped("product_id").ids if
+                                 p not in contract.contract_line_ids.mapped("product_id").ids]
+        if diff_product_invoices:
+            for diff_product_inv in diff_product_invoices:
+                res.append((2, line_ids.filtered(lambda l: l.product_id.id == diff_product_inv).id, 0))
+        # Line to create in the invoice
+        # If the contract has different product than the invoice we should create the line
+        diff_product_contract = [p for p in contract.contract_line_ids.mapped("product_id").ids if
+                                 p not in line_ids.mapped("product_id").ids]
+        if diff_product_contract:
+            for diff_product_contr in diff_product_contract:
+                contract_line = contract.contract_line_ids.filtered(lambda l: l.product_id.id == diff_product_contr)
+                if contract_line:
+                    res.append((0, 0,
+                                {
+                                    "price_unit": contract_line.amount,
+                                    "product_id": contract_line.product_id.id,
+                                    'name': contract_line.product_id.name,
+                                    'quantity': contract_line.quantity,
+                                    'contract_id': contract_line.contract_id.id,
+                                    'account_id': contract_line.product_id.with_company(
+                                        contract_line.contract_id.company_id.id).property_account_income_id.id or False
+                                }
+                                ))
```

Comment:
Instead of coding the dictionary here, we could maybe move the method `def get_inv_lines_data(self):` declared in `recurring.contract` into the `recurring.contract.line` object instead so that we could call the method here with all the missing lines at once.
```suggestion
        missing_lines = contract.contract_line_ids.filtered(
            lambda l: l.product_id not in line_ids.mapped("product_id"))
        res.extend([(0, 0, line_vals) for line_vals in missing_lines.get_inv_lines_data()])
```